### DISCUSSION TITLE
Migrating to 2.0. Error

### DISCUSSION THREAD

**Discussion by Patrick386** [firebase/flutterfire#6061] [Q&A]
Migrating to 2.0.
Below is the error code. What is the solution?

There is too much code to fix.
I need a variety of examples and explanations for the migration.
What can you get through migration?

```
 _CodeRef = CodeCollectionRef.doc();
   
      _itemCode = _itemCode.copyWith(
          userId: _authUser.uid,
          packageDocId: userPackagesRef.id,
          productDocId: productRef.id,
          Code: _CodeRef.id, 
          savedDateAt: 0,
          transferDateAt: 0);

      batch.set<Map<String, dynamic>>(_CodeRef, _itemQrCode.toJson());
```

**Error Message**
`Expected a value of type 'DocumentReference<Map<String, dynamic>>', but got one of type '_WithConverterDocumentReference<CodeModel>'`

**Comment by stargazing-dino** [comment]
The point of the v2 is to have type safe conversions from Firebase types to custom types. Type safety helps you ensure the documents you pull from firestore are actually in the shape you except which can help you avoid a lot of runtime issues by forcing you to deal with them at compile time.

As for your problem, I think it's because `batch.set` is now over a type T. Here is the function signature

```dart
void set<T>(
  DocumentReference<T> document,
  T data, [
  SetOptions? options,
])
```

Which with how you call it you're saying you want a `DocumentReference<Map<String, dynamic>>`. However, _CodeRef is a type `DocumentReference<CodeModel>`. It's just a type mismatch.

Just call `batch.set(_CodeRef, _itemQrCode.toJson())` or `batch.set<CodeModel>(_CodeRef, _itemQrCode.toJson())`

**Reply by Patrick386** [reply]
I tried again, but it's the same error.
I'm using the Freezed, JsonSerializable package and it's up to date.
It's hard to find out what's wrong.

Freezed
JsonSerializable
```
 factory CodeModel.formJson(Map<String, Object?> json) =>
      _$CodeModelFromJson(json);

  Map<String, dynamic> toJson() => _$CodeModelToJson(this);
```

DB Ref Service

`final CollectionReference<CodeModel> CodeCollectionRef = firebaseFirestore
    .collection(CODE_COLLECTION)
    .withConverter<CodeModel>(
        fromFirestore: (DocumentSnapshot<Map<String, dynamic>> snapshot, _) =>
            CodeModel.formJson(snapshot.data()!),
        toFirestore: (CodeModel model, _) => model.toJson());`
